code/datums/components/crafting/guncrafting.dm

```diff
@@ -2,20 +2,130 @@
+	icon = 'icons/obj/improvised.dmi'
+
+/obj/item/weaponcrafting/silkstring
+	name = "silkstring"
+	desc = "A long pice of silk looks like cable coil."
+	icon = 'icons/obj/improvised.dmi'
+	icon_state = "silkstring"
+
 /obj/item/weaponcrafting/receiver
 	name = "modular receiver"
 	desc = "A prototype modular receiver and trigger assembly for a firearm."
-	icon = 'icons/obj/improvised.dmi'
 	icon_state = "receiver"
 
 /obj/item/weaponcrafting/stock
 	name = "rifle stock"
 	desc = "A classic rifle stock that doubles as a grip, roughly carved out of wood."
-	icon = 'icons/obj/improvised.dmi'
 	icon_state = "riflestock"
 
-/obj/item/weaponcrafting/silkstring
-	name = "silkstring"
-	desc = "A long pice of silk looks like cable coil."
-	icon = 'icons/obj/improvised.dmi'
-	icon_state = "silkstring"
\ No newline at end of file
+/obj/item/weaponcrafting/stock/attackby(obj/item/S, mob/user, params)
+	..()
+	if(istype(S, /obj/item/pipe))
+		return
+	if(istype(S, /obj/item/weaponcrafting/receiver))
+		var/obj/item/weaponcrafting/receiver/stock/R = /obj/item/weaponcrafting/receiver/stock
+		to_chat(user, "<span class='notice'>You add [S] to [src]!</span>")
+		qdel(S)
+		qdel(src)
+		new R(user.loc, 1)
+
+/obj/item/weaponcrafting/receiver/stock
+	name = "receiver stock"
+	desc = "A classic rifle stock that doubles as a grip, roughly carved out of wood, this one has a modular receiver attached."
+	icon_state = "mounted_receiver"
+
+/obj/item/weaponcrafting/receiver/stock/attackby(obj/item/P, mob/user, params)
+	..()
+	if(istype(P, /obj/item/pipe))
+		var/obj/item/weaponcrafting/receiver/stock/barreled/B = /obj/item/weaponcrafting/receiver/stock/barreled
+		to_chat(user, "<span class='notice'>You add [P] to [src]!</span>")
+		qdel(P)
+		qdel(src)
+		new B(user.loc, 1)
+
+/obj/item/weaponcrafting/receiver/stock/barreled
+	name = "gun assembly"
+	desc = "A classic rifle stock that doubles as a grip, roughly carved out of wood. This one has a modular receiver attached."
+	icon_state = "mounted_receiver_piped"
+
+/obj/item/weaponcrafting/receiver/stock/barrled/attackby(obj/item/W, mob/user, params)
+	..()
+	if(istype(W, /obj/item/stack/packageWrap))
+		var/obj/item/weaponcrafting/receiver/stock/barreled/gun/G = /obj/item/weaponcrafting/receiver/stock/barreled/gun
+		if(W.use(15))
+			to_chat(user, "<span class='notice'>You add [W] to [src]!</span>")
+			qdel(src)
+			new G(user.loc, 1)
+		else
+			to_chat(user, "<span class='warning'>You need at least ten lengths of package wrap if you want to attach this pipe!</span>")
+
+/obj/item/weaponcrafting/receiver/stock/barreled/gun
+	name = "gun assembly"
+	desc = "A classic rifle stock that doubles as a grip, roughly carved out of wood. This one has a modular receiver firmly attached. Adding a new pipe would allow for a rifle to be made."
+	icon_state = "mounted_receiver_piped_wrapped"
+
+/obj/item/weaponcrafting/receiver/stock/barreled/gun/attackby(obj/item/D, mob/user, params)
+	if(istype(D, /obj/item/screwdriver))
```

Comment:
Use tool behaviours checks or the appropriate tools act procedure if possible, not type checks.
This also applies to all other steps.